paraatm/simulation_method/aviationr_model.py

```diff
@@ -0,0 +1,372 @@
+"""
+NASA NextGen NAS ULI Information Fusion
+
+@organization: Arizona State University
+@author: Xinyu Zhao
+@date: 2020-05-30
+@last updated: 2020-05-30
+
+This Python script is used for simulating aviation accident based on recording in NTSB
+Simulation based on NATS beta1.7 standalone version
+"""
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import pandas as pd
+import numpy as np
+
+class RiskEstimator:
+    """
+
+        Estimating the risk of each event in an accident
+
+        Parameters
+        ----------
+        inputs to Aviation Risk simulation
+        isRNN : boolean
+            check weather using RNN or not
+        data : str
+            directory to required reference from NTSB
+        model : str
+            directory to pre-trained model
+        device : torch.device
+            'CUDA'/ 'CPU'-- Running the model along GPU or CPU
+    """
+    def __init__(self, model, isRNN, data, device):
+        self.model = model
+        self.isRNN = isRNN
+        self.data = data
+        if device is None:
+            device = torch.device('cpu')
+            if torch.cuda.is_available():
+                device = torch.device('cuda')
+        self.device = device
+
+    def load_model(self):
+        """
+            Load pre-trained models
+        """
+        refer = pd.read_pickle(self.data + 'refer.pickle')
+        occurrence_code_corpus = refer['occurrence_code_corpus']
+        subject_code_corpus = refer['subject_code_corpus']
+        phase_code_corpus = refer['phase_code_corpus']
+        subj_to_occurrence_dict = refer['subj_to_occurrence_dict']
+        occurrence_to_subj_dict = refer['occurrence_to_subj_dict']
+        phase_to_occurrence_dict = refer['phase_to_occurrence_dict']
+        subject_size = len(subject_code_corpus)
+        phase_size = len(phase_code_corpus)
+        occurrence_size = len(occurrence_code_corpus)
+
+        batch_size = 1
+        try:
+            import torch
+        except ImportError:
+            print('aviation risk module requires installation of torch')
+            raise
+        if self.isRNN:
+            embedding_dim = 30  # Embedding of each word
+            hidden_dim = 50  # Final hidden dimension that is used for prediction
+            num_layer = 2  # Number of layers
+            latent_dim = 30  # Latend dimension
+
+            model = _RNNModel(phase_size, occurrence_size, subject_size, embedding_dim, latent_dim, num_layer,
+                             batch_size, hidden_dim, self.device).to(self.device)
+            hierarchical_softmax = _HierarchicalSoftmax(hidden_dim, phase_to_occurrence_dict, occurrence_to_subj_dict,
+                                                       subj_to_occurrence_dict, self.device).to(self.device)
+            hierarchical_softmax.load_state_dict(torch.load(self.model + "hierarchical_softmax_rnn.sav"))
```

Comment:
With some trial and error (such as using a delay value of 30 instead of 100), I am occasionally able to get the test to find the accident (however, on my computer, it is not repeatable -- sometimes it finds the accident, sometimes it does not).  When it does find the accident, I get the following error, which seems to originate from this `torch.load` call:
```
Traceback (most recent call last):
  File "paraatm/tests/aviationr_example.py", line 18, in <module>
    result  = sim.simulation()  # call simulation function using NatsSimulationWrapper
  File "/home/jmcfarland/work/para-atm/paraatm/simulation_method/aviationr.py", line 139, in simulation
    model, hierarchical_softmax, risk_model = risk_estimator.load_model()
  File "/home/jmcfarland/work/para-atm/paraatm/simulation_method/aviationr_model.py", line 77, in load_model
    hierarchical_softmax.load_state_dict(torch.load(self.model + "hierarchical_softmax_rnn.sav"))
  File "/home/jmcfarland/venv/para-xz/lib/python3.8/site-packages/torch/serialization.py", line 593, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
  File "/home/jmcfarland/venv/para-xz/lib/python3.8/site-packages/torch/serialization.py", line 773, in _legacy_load
    result = unpickler.load()
  File "/home/jmcfarland/venv/para-xz/lib/python3.8/site-packages/torch/serialization.py", line 729, in persistent_load
    deserialized_objects[root_key] = restore_location(obj, location)
  File "/home/jmcfarland/venv/para-xz/lib/python3.8/site-packages/torch/serialization.py", line 178, in default_restore_location
    result = fn(storage, location)
  File "/home/jmcfarland/venv/para-xz/lib/python3.8/site-packages/torch/serialization.py", line 154, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "/home/jmcfarland/venv/para-xz/lib/python3.8/site-packages/torch/serialization.py", line 138, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
```

Reply:
I have modified the code a little bit. It should work for machine with CPU now. Please take a look.

Reply:
Thanks, it works now.